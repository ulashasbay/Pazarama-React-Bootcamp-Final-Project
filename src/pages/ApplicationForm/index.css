.application-form-container {
  font-family: "Open Sans", sans-serif;
  color: #3a3c47;
  line-height: 1.6;
  display: flex;
  flex-direction: column;
  align-items: center;
  margin: 0;
  padding: 0;
  height: auto;
}

.application-form-container h1 {
  margin-top: 28px;
  font-size: 30px;
}

.application-form {
  background: #fff;
  max-width: 500px;
  width: 100%;
  padding: 58px 44px;
  border: 1px solid #e1e2f0;
  border-radius: 4px;
  box-shadow: 0 0 5px 0 rgba(42, 45, 48, 0.12);
  transition: all 0.3s ease;
  margin-bottom: 5px;
}

.application-form-row {
  display: flex;
  flex-direction: column;
  margin-bottom: 20px;
}

.application-form-row label {
  font-size: 13px;
  color: #8086a9;
}

.application-form-row input {
  flex: 1;
  padding: 13px;
  border: 1px solid #d6d8e6;
  border-radius: 4px;
  font-size: 16px;
  transition: all 0.2s ease-out;
}

.application-form-row input:focus {
  outline: none;
  box-shadow: inset 2px 2px 5px 0 rgba(42, 45, 48, 0.12);
}

.application-form-row input::placeholder {
  color: #c8cddf;
}

/* text area alanları için */
.application-form-row textarea {
  flex: 1;
  padding: 13px;
  border: 1px solid #d6d8e6;
  border-radius: 4px;
  font-size: 16px;
  transition: all 0.2s ease-out;
}

.application-form-row textarea:focus {
  outline: none;
  box-shadow: inset 2px 2px 5px 0 rgba(42, 45, 48, 0.12);
}

.application-form-row textarea::placeholder {
  color: #c8cddf;
}

.login-button {
  width: 100%;
  padding: 12px;
  font-size: 18px;
  background: #ff008b;
  color: #fff;
  border: none;
  border-radius: 100px;
  cursor: pointer;
  font-family: "Open Sans", sans-serif;
  margin-top: 15px;
}

.login-button:hover {
  background: #fd3ea7;
}

.app-form-errors {
  margin-top: 0;
  font-size: 14px;
  color: red;
}
.file-upload-progress {
  margin-top: 0;
  font-size: 14px;
  color: rgb(0, 0, 0);
}

.form-loading {
  display: flex;
  justify-content: center;
  align-items: center;
  text-align: center;
  margin-top: 14rem;
}

@media (max-width: 960px) {
  .application-form-container {
    margin: 0 18px;
  }

  .application-form {
    border: none;
    box-shadow: none;
    padding: 20px 0;
  }
}
