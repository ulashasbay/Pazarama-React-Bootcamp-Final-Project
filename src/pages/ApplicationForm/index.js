import React, { useState } from "react";
import { useFormik } from "formik";
import Input from "../../components/Input";
import FormSubmitBtn from "../../components/FormSubmitBtn";
import TextArea from "../../components/TextArea";
import ApplicationFormVal from "../../schema/ApplicationFormVal";
import { collection, addDoc } from "firebase/firestore";
import { db, storage } from "../../config/firebase";
import { ref, getDownloadURL, uploadBytesResumable } from "firebase/storage";
import { useNavigate } from "react-router-dom";
import { useDispatch } from "react-redux";
import { updateAppId } from "../../redux/appId/appIdSlice";
import { updateUserInfo } from "../../redux/userInfo/userInfoSlice";
import "./index.css";

function ApplicationForm() {
  let navigate = useNavigate();
  const dispatch = useDispatch();

  const [isLoading, setIsLoading] = useState(false);
  const [uploadProgres, setUploadProgress] = useState("");

  // Başvuru formuna kullanıcı tarafından fotograf eklendiğinde form submit edilirken çalışan fonksiyon
  // - Fotografı firebase storage a upload ediyor
  // - Upload edilen Fotografın url sini alıp kullanıcı başvuru bilgilerine ekliyor
  // - Kullanıcı başvuru bilgilerini firebase database e upload ediyor
  const createApplicationWithImg = (values) => {
    const file = values.imageUrl;
    const storageRef = ref(storage, `files/${file.name}`);
    const uploadTask = uploadBytesResumable(storageRef, file);
    uploadTask.on(
      "state_changed",
      (snapshot) => {
        const prog = Math.round(
          (snapshot.bytesTransferred / snapshot.totalBytes) * 100
        );
        setIsLoading(true);
        setUploadProgress(prog);
      },
      (error) => console.log(error),
      () => {
        getDownloadURL(uploadTask.snapshot.ref).then((downloadURL) => {
          values.imageUrl = downloadURL;
          createApplication(values);
        });
      }
    );
  };

  // Başvuru formuna kullanıcı tarafından fotograf eklenmediğinde form submit edilirken çalışan fonksiyon
  // Kullanıcı başvuru bilgilerini Firebase Database e upload ediyor
  const userColRef = collection(db, "applications");
  const createApplication = async (data) => {
    const docRef = await addDoc(userColRef, data);
    dispatch(updateAppId(docRef.id));
    dispatch(updateUserInfo(data));
    navigate(`/basvuru-basarili`);
    setIsLoading(false);
  };

  const {
    handleSubmit,
    handleChange,
    handleBlur,
    errors,
    touched,
    setFieldValue,
  } = useFormik({
    initialValues: {
      name: "",
      surname: "",
      age: "",
      tcNo: "",
      applicationReason: "",
      address: "",
      applicationStatus: "Bekliyor",
      applicationResult: "Belirsiz",
      applicationDate: new Date()
        .toLocaleString("en-GB", {
          month: "2-digit",
          day: "2-digit",
          year: "numeric",
        })
        .slice(0, 10),
    },
    onSubmit: (values) => {
      if (!("imageUrl" in values)) {
        createApplication(values);
      } else {
        createApplicationWithImg(values);
      }
    },
    validationSchema: ApplicationFormVal,
  });
  return (
    <>
      {isLoading ? (
        <div className="form-loading">Gönderiliyor... %{uploadProgres}</div>
      ) : (
        <div className="application-form-container">
          <h1>Başvuru Formu</h1>
          <form className="application-form" onSubmit={handleSubmit}>
            <div className="application-form-row">
              <Input
                name="name"
                type="text"
                text="Ad"
                placeholder="Ad"
                onChange={handleChange}
                onBlur={handleBlur}
              />
              {errors.name && touched.name && (
                <span className="app-form-errors">{errors.name}</span>
              )}
            </div>
            <div className="application-form-row">
              <Input
                name="surname"
                type="text"
                text="Soyad"
                placeholder="Soyad"
                onChange={handleChange}
                onBlur={handleBlur}
              />
              {errors.surname && touched.surname && (
                <span className="app-form-errors">{errors.surname}</span>
              )}
            </div>
            <div className="application-form-row">
              <Input
                name="age"
                type="text"
                text="Yaş"
                placeholder="Yaş"
                onChange={handleChange}
                onBlur={handleBlur}
              />
              {errors.age && touched.age && (
                <span className="app-form-errors">{errors.age}</span>
              )}
            </div>
            <div className="application-form-row">
              <Input
                name="tcNo"
                type="text"
                text="T.C. Kimlik No"
                placeholder="T.C. Kimlik No"
                onChange={handleChange}
                onBlur={handleBlur}
              />
              {errors.tcNo && touched.tcNo && (
                <span className="app-form-errors">{errors.tcNo}</span>
              )}
            </div>
            <div className="application-form-row">
              <Input
                name="applicationReason"
                type="text"
                text="Başvuru Nedeni"
                placeholder="Başvuru Nedeni"
                onChange={handleChange}
                onBlur={handleBlur}
              />
              {errors.applicationReason && touched.applicationReason && (
                <span className="app-form-errors">
                  {errors.applicationReason}
                </span>
              )}
            </div>
            <div className="application-form-row">
              <TextArea
                text="Adres"
                name="address"
                placeholder="Adres"
                onChange={handleChange}
                onBlur={handleBlur}
              />
              {errors.address && touched.address && (
                <span className="app-form-errors">{errors.address}</span>
              )}
            </div>
            <div className="application-form-row">
              <Input
                name="imageUrl"
                type="file"
                text="Fotoğraf"
                placeholder="Fotoğraf"
                onChange={(event) => {
                  setFieldValue("imageUrl", event.currentTarget.files[0]);
                }}
                onBlur={handleBlur}
              />
              {errors.imageUrl && touched.imageUrl && (
                <span className="app-form-errors">{errors.imageUrl}</span>
              )}
            </div>
            <FormSubmitBtn disabled={isLoading} text="Gönder" />
          </form>
        </div>
      )}
    </>
  );
}

export default ApplicationForm;
