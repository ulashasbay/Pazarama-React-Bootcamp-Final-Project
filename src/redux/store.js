import { configureStore } from "@reduxjs/toolkit";
import appIdReducer from "./appId/appIdSlice";
import userInfoReducer from "./userInfo/userInfoSlice";
import isLoggedInReducer from "./isLoggedIn/isLoggedInSlice";

export const store = configureStore({
  reducer: {
    appId: appIdReducer,
    userInfo: userInfoReducer,
    isLoggedIn: isLoggedInReducer,
  },
});
