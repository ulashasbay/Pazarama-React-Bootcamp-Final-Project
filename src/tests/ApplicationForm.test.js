import React from "react";
import { render, screen } from "@testing-library/react";
import ApplicationForm from "../pages/ApplicationForm";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";
import { store } from "../redux/store";
import renderer from "react-test-renderer";

const AllTheProviders = ({ children }) => {
  return (
    <BrowserRouter>
      <Provider store={store}>{children}</Provider>
    </BrowserRouter>
  );
};

describe("ApplicationForm Tests", () => {
  beforeEach(() => {
    render(<ApplicationForm />, { wrapper: AllTheProviders });
  });

  test("Render Header", () => {
    const header = screen.getByText("Başvuru Formu");
    expect(header).toBeInTheDocument();
  });

  test("Render Ad Label", () => {
    const adLabel = screen.getByText("Ad");
    expect(adLabel).toBeInTheDocument();
  });

  test("Render Soyad Label", () => {
    const soyadLabel = screen.getByText("Soyad");
    expect(soyadLabel).toBeInTheDocument();
  });

  test("Render Yaş Label", () => {
    const yasLabel = screen.getByText("Yaş");
    expect(yasLabel).toBeInTheDocument();
  });

  test("Render tcNo Label", () => {
    const tcNoLabel = screen.getByText("T.C. Kimlik No");
    expect(tcNoLabel).toBeInTheDocument();
  });

  test("Render Başvuru Nedeni Label", () => {
    const basvuruNedeniLabel = screen.getByText("Başvuru Nedeni");
    expect(basvuruNedeniLabel).toBeInTheDocument();
  });

  test("Render Adres Label", () => {
    const adresLabel = screen.getByText("Adres");
    expect(adresLabel).toBeInTheDocument();
  });

  test("Render Fotograf Label", () => {
    const fotoLabel = screen.getByText("Fotoğraf");
    expect(fotoLabel).toBeInTheDocument();
  });

  test("Ad input type is text", () => {
    const adInput = screen.getByPlaceholderText("Ad");
    expect(adInput.type).toEqual("text");
  });

  test("Soyad input type is text", () => {
    const soyadInput = screen.getByPlaceholderText("Soyad");
    expect(soyadInput.type).toEqual("text");
  });

  test("Yaş input type is text", () => {
    const yasInput = screen.getByPlaceholderText("Yaş");
    expect(yasInput.type).toEqual("text");
  });

  test("tcNo input type is text", () => {
    const tcNoInput = screen.getByPlaceholderText("T.C. Kimlik No");
    expect(tcNoInput.type).toEqual("text");
  });

  test("Başvuru Nedeni input type is text", () => {
    const basvuruNedeniInput = screen.getByPlaceholderText("Başvuru Nedeni");
    expect(basvuruNedeniInput.type).toEqual("text");
  });

  test("Gönder button type is submit", () => {
    const button = screen.getByText("Gönder");
    expect(button.type).toEqual("submit");
  });
});

test("ApplicationForm renders correctly", () => {
  const tree = renderer
    .create(
      <BrowserRouter>
        <Provider store={store}>
          <ApplicationForm />
        </Provider>
      </BrowserRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
