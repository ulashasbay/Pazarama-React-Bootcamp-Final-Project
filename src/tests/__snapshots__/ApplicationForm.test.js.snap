// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationForm renders correctly 1`] = `
<div
  className="application-form-container"
>
  <h1>
    Başvuru Formu
  </h1>
  <form
    className="application-form"
    onSubmit={[Function]}
  >
    <div
      className="application-form-row"
    >
      <label
        htmlFor="name"
      >
        Ad
      </label>
      <input
        name="name"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Ad"
        type="text"
      />
    </div>
    <div
      className="application-form-row"
    >
      <label
        htmlFor="surname"
      >
        Soyad
      </label>
      <input
        name="surname"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Soyad"
        type="text"
      />
    </div>
    <div
      className="application-form-row"
    >
      <label
        htmlFor="age"
      >
        Yaş
      </label>
      <input
        name="age"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Yaş"
        type="text"
      />
    </div>
    <div
      className="application-form-row"
    >
      <label
        htmlFor="tcNo"
      >
        T.C. Kimlik No
      </label>
      <input
        name="tcNo"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="T.C. Kimlik No"
        type="text"
      />
    </div>
    <div
      className="application-form-row"
    >
      <label
        htmlFor="applicationReason"
      >
        Başvuru Nedeni
      </label>
      <input
        name="applicationReason"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Başvuru Nedeni"
        type="text"
      />
    </div>
    <div
      className="application-form-row"
    >
      <label
        htmlFor="address"
      >
        Adres
      </label>
      <textarea
        name="address"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Adres"
      />
    </div>
    <div
      className="application-form-row"
    >
      <label
        htmlFor="imageUrl"
      >
        Fotoğraf
      </label>
      <input
        name="imageUrl"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Fotoğraf"
        type="file"
      />
    </div>
    <button
      className="app-check-button"
      disabled={false}
      type="submit"
    >
      Gönder
    </button>
  </form>
</div>
`;
